cmd/nfd-topology-updater/main.go

```diff
@@ -0,0 +1,195 @@
+*/
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/docopt/docopt-go"
+	v1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+	"sigs.k8s.io/node-feature-discovery/pkg/dumpobject"
+	"sigs.k8s.io/node-feature-discovery/pkg/kubeconf"
+	topology "sigs.k8s.io/node-feature-discovery/pkg/nfd-topology-updater"
+	"sigs.k8s.io/node-feature-discovery/pkg/podres"
+	"sigs.k8s.io/node-feature-discovery/pkg/resourcemonitor"
+	"sigs.k8s.io/node-feature-discovery/pkg/version"
+)
+
+const (
+	// ProgramName is the canonical name of this program
+	ProgramName = "nfd-topology-updater"
+)
+
+func main() {
+	// Assert that the version is known
+	if version.Undefined() {
+		log.Printf("WARNING: version not set! Set -ldflags \"-X sigs.k8s.io/node-feature-discovery/pkg/version.version=`git describe --tags --dirty --always`\" during build or run.")
+	}
+
+	args, resourcemonitorArgs, err := argsParse(nil)
+	if err != nil {
+		log.Fatalf("failed to parse command line: %v", err)
+	}
+
+	klConfig, err := kubeconf.GetKubeletConfigFromLocalFile(resourcemonitorArgs.KubeletConfigFile)
+	if err != nil {
+		log.Fatalf("error getting topology Manager Policy: %v", err)
+	}
+	tmPolicy := klConfig.TopologyManagerPolicy
+	log.Printf("Detected kubelet Topology Manager policy %q", tmPolicy)
+
+	podResClient, err := podres.GetPodResClient(resourcemonitorArgs.PodResourceSocketPath)
+	if err != nil {
+		log.Fatalf("Failed to get PodResource Client: %v", err)
+	}
+	var resScan resourcemonitor.ResourcesScanner
+
+	resScan, err = resourcemonitor.NewPodResourcesScanner(resourcemonitorArgs.Namespace, podResClient)
+	if err != nil {
+		log.Fatalf("Failed to initialize ResourceMonitor instance: %v", err)
+	}
+
+	// CAUTION: these resources are expected to change rarely - if ever.
+	// So we are intentionally do this once during the process lifecycle.
+	// TODO: Obtain node resources dynamically from the podresource API
+	zonesChannel := make(chan v1alpha1.ZoneList)
+	var zones v1alpha1.ZoneList
+
+	resAggr, err := resourcemonitor.NewResourcesAggregator(resourcemonitorArgs.SysfsRoot, podResClient)
+	if err != nil {
+		log.Fatalf("Failed to obtain node resource information: %v", err)
+	}
+
+	log.Printf("resAggr is: %v\n", resAggr)
+	go func() {
+		for {
+			log.Printf("Scanning\n")
+
+			podResources, err := resScan.Scan()
+			log.Printf("podResources are: %v\n", podResources)
+			if err != nil {
+				log.Printf("Scan failed: %v\n", err)
+				continue
+			}
+
+			zones = resAggr.Aggregate(podResources)
+			zonesChannel <- zones
+			log.Printf("After aggregating resources identified zones are:%v", dumpobject.DumpObject(zones))
+
+			time.Sleep(resourcemonitorArgs.SleepInterval)
+		}
+	}()
+
+	// Get new TopologyUpdater instance
+	instance, err := topology.NewTopologyUpdater(args, tmPolicy)
+	if err != nil {
+		log.Fatalf("Failed to initialize NfdWorker instance: %v", err)
+	}
+	for {
+
+		zonesValue := <-zonesChannel
```

Comment:
The idea was to separate the logic of resource scanning that happens every sleep interval from logic responsible for resource aggregation (followed by sending the information to the nfd-master). But you do make a good point this can  easily be done a single loop. Will do that.